src/common/walFilter/versions/recordProcessGPDB6.c

```diff
@@ -0,0 +1,620 @@
+#define XLOG_SPGIST_VACUUM_ROOT     0x70
+#define XLOG_SPGIST_VACUUM_REDIRECT 0x80
+
+// bitmap
+#define XLOG_BITMAP_INSERT_LOVITEM  0x20 /* add a new entry into a LOV page */
+#define XLOG_BITMAP_INSERT_META     0x30 /* update the metapage */
+#define XLOG_BITMAP_INSERT_BITMAP_LASTWORDS 0x40 /* update the last 2 words
+                                                    in a bitmap */
+/* insert bitmap words into a bitmap page which is not the last one. */
+#define XLOG_BITMAP_INSERT_WORDS        0x50
+/* 0x60 is unused */
+#define XLOG_BITMAP_UPDATEWORD          0x70
+#define XLOG_BITMAP_UPDATEWORDS         0x80
+
+// appendonly
+#define XLOG_APPENDONLY_INSERT          0x00
+#define XLOG_APPENDONLY_TRUNCATE        0x10
+
+typedef uint32_t CommandId;
+typedef uint16_t OffsetNumber;
+
+typedef enum ForkNumber
+{
+    InvalidForkNumber = -1,
+    MAIN_FORKNUM = 0,
+    FSM_FORKNUM,
+    VISIBILITYMAP_FORKNUM,
+
+    /*
+     * Init forks are used to create an initial state that can be used to
+     * quickly revert an object back to its empty state. This is useful for
+     * reverting unlogged tables and indexes back to their initial state during
+     * recovery.
+     */
+    INIT_FORKNUM
+
+    /*
+     * NOTE: if you add a new fork, change MAX_FORKNUM and possibly
+     * FORKNAMECHARS below, and update the forkNames array in
+     * src/common/relpath.c
+     */
+} ForkNumber;
+
+typedef struct BkpBlock
+{
+    RelFileNode node;           /* relation containing block */
+    ForkNumber fork;            /* fork within the relation */
+    BlockNumber block;          /* block number */
+    uint16_t hole_offset;         /* number of bytes before "hole" */
+    uint16_t hole_length;         /* number of bytes in "hole" */
+
+    /* ACTUAL BLOCK DATA FOLLOWS AT END OF STRUCT */
+} BkpBlock;
+
+typedef struct xl_smgr_truncate
+{
+    BlockNumber blkno;
+    RelFileNode rnode;
+} xl_smgr_truncate;
+
+typedef struct BlockIdData
+{
+    uint16_t bi_hi;
+    uint16_t bi_lo;
+} BlockIdData;
+
+typedef struct ItemPointerData
+{
+    BlockIdData ip_blkid;
+    OffsetNumber ip_posid;
+}
+
+#ifdef __arm__
+__attribute__((packed))         /* Appropriate whack upside the head for ARM */
+#endif
+ItemPointerData;
+
+typedef struct xl_heaptid
+{
+    RelFileNode node;
+    ItemPointerData tid;        /* changed tuple id */
+} xl_heaptid;
+
+typedef struct xl_heap_new_cid
+{
+    /*
+     * store toplevel xid so we don't have to merge cids from different
+     * transactions
+     */
+    TransactionId top_xid;
+    CommandId cmin;
+    CommandId cmax;
+
+    /*
+     * don't really need the combocid since we have the actual values right in
+     * this struct, but the padding makes it free and its useful for
+     * debugging.
+     */
+    CommandId combocid;
+
+    /*
+     * Store the relfilenode/ctid pair to facilitate lookups.
+     */
+    xl_heaptid target;
+} xl_heap_new_cid;
+
+static
+bool
+getXlog(const XLogRecord *record, RelFileNode **node)
```

Comment:
Could you please add comments to the functions below, which one does what, why certain values ​​are ignored, etc.?